File: src/main/java/discretemath/bool/operator/Operator.java
package discretemath.bool.operator;

import discretemath.bool.expression.BooleanExpression;
import discretemath.bool.expression.exception.InvalidArgumentCountException;

public interface Operator {

	boolean evaluate(boolean... values);

	BooleanExpression applyTo(BooleanExpression... operands) throws InvalidArgumentCountException;

	BooleanExpression applyUsing(FunctionallyCompleteOperatorsSet operatorsSet,
								 BooleanExpression... operands) throws InvalidArgumentCountException, UnsupportedOperationException;

	String toString(BooleanExpression... operands) throws InvalidArgumentCountException;
}

File: src/main/scala/discretemath/bool/operator/Operator.scala
package discretemath.bool.operator

import discretemath.bool.expression.BooleanExpression;


trait Operator {

  def evaluate(values: Seq[Boolean]): Boolean

  final def %(values: Boolean*): Boolean = evaluate(values)

  def inject(operands: Seq[BooleanExpression]): BooleanExpression

  @throws[UnsupportedOperationException]
  def injectUsing(operatorsSet: FunctionallyCompleteOperatorsSet, operands: Seq[BooleanExpression]): BooleanExpression

  def toString(operands: Seq[BooleanExpression]): String
}